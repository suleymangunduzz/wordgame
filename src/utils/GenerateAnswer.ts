const generateAnswer = (selectedNames: string[], names: string[]) => {
  const lastValidAnswer = selectedNames[selectedNames.length - 1];

  const lastValidAnswerEndsWith = lastValidAnswer.charAt(lastValidAnswer.length - 1);

  const unusedNameList = names.filter(name => !selectedNames.includes(name));

  const newAnswer = unusedNameList.find(name => name.charAt(0) === lastValidAnswerEndsWith);

  return newAnswer || '';
};

export default generateAnswer;
