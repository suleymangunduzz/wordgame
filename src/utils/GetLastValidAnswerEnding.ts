const getLastValidAnswerEnding = (selectedNames: string[]) => {
  const lastValidAnswer = selectedNames[selectedNames.length - 1];

  return lastValidAnswer.charAt(lastValidAnswer.length - 1);
};

export default getLastValidAnswerEnding;
