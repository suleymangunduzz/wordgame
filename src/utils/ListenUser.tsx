import checkAnswer from './CheckAnswer';

export const listenUser = (
  recognition: any,
  selectedNames: string[],
  names: string[],
  setWinner: (name: string) => void,
  setTimeLeft: (timeLeft: number) => void,
  setGameStatus: (status: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setUsersFinalAnswer: (answer: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
) => {
  console.log("Well, it is user's turn, lets play it !");

  recognition.start();

  // eslint-disable-next-line no-param-reassign
  recognition.onresult = (event: any) => {
    let finalTranscript = '';

    for (let i = event.resultIndex; i < event.results.length; i += 1) {
      const { transcript } = event.results[i][0];
      if (event.results[i].isFinal) finalTranscript += `${transcript} `;
    }
    if (event.results[0].isFinal) {
      console.log("User's answer: ", finalTranscript.toLocaleLowerCase('tr'));
      setUsersFinalAnswer(finalTranscript.toLocaleLowerCase('tr'));
      checkAnswer(
        finalTranscript.toLocaleLowerCase('tr'),
        selectedNames,
        names,
        setGameStatus,
        setNextPlayer,
        setWinner,
        setSelectedNames,
        setTimeLeft,
      );
    }
  };
};

export default listenUser;
