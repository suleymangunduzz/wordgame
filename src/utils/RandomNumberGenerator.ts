const generateRandomNumber = (maxNumber: number) => {
  return Math.floor(Math.random() * Math.floor(maxNumber)) + 1;
};

export default generateRandomNumber;
