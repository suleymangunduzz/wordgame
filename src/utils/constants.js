export const FINISHED = 'FINISHED';
export const NOT_STARTED = 'NOT_STARTED';
export const STARTED = 'STARTED';

export const USER = 'USER';
export const MACHINE = 'MACHINE';
