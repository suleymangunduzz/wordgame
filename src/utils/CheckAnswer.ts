import getLastValidAnswerEnding from './GetLastValidAnswerEnding';
import { MACHINE, FINISHED } from './constants';

export const checkAnswer = (
  answer: string,
  selectedNames: string[],
  names: string[],
  setGameStatus: (gameStatus: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setWinner: (winner: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
  setTimeLeft: (timeLeft: number) => any,
) => {
  const cleanAnswer = answer.replace(/\s+/g, '');
  const answerStartsWith = cleanAnswer.charAt(0);

  const lastValidAnswerEndsWith = getLastValidAnswerEnding(selectedNames);

  if (selectedNames.some(name => name === cleanAnswer)) {
    console.log('The answer is used before, so the game is finishing.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  if (answerStartsWith !== lastValidAnswerEndsWith) {
    console.log('The answer does not match with the last valid answer.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  if (!names.includes(cleanAnswer)) {
    console.log('There is no name like this in our data.');
    setGameStatus(FINISHED);
    setNextPlayer(MACHINE);
    setWinner(MACHINE);
    return;
  }

  console.log('The answer is correct and added to the our list, good job, keep playing!');
  const newState: string[] = [...selectedNames, cleanAnswer];
  setSelectedNames(newState);
  setNextPlayer(MACHINE);
  setTimeLeft(8);
};

export default checkAnswer;
