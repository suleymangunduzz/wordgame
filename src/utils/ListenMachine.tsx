import generateAnswer from './GenerateAnswer';
import { USER, FINISHED } from './constants';

export const listenMachine = (
  isLucky: boolean,
  timeToWait: number,
  selectedNames: string[],
  names: string[],
  synthRef: any,
  setWinner: (name: string) => void,
  setTimeLeft: (timeLeft: number) => void,
  setGameStatus: (status: string) => void,
  setNextPlayer: (nextPlayer: string) => void,
  setUsersFinalAnswer: (answer: string) => void,
  setSelectedNames: (selectedNames: string[]) => void,
  isFirtTime?: boolean,
) => {
  console.log('Machine is playing...');
  setUsersFinalAnswer('');

  if (isLucky) {
    console.log('Luckily, it keeps playing...');

    console.log(`After ${timeToWait} seconds machine is gonna answer.`);

    let newSelection = '';
    let newState: string[] = [];

    if (isFirtTime) {
      newSelection = names[timeToWait];
      newState = [newSelection];
    } else {
      console.log('Generate some valid name for the machine if it is possible.');
      newSelection = generateAnswer(selectedNames, names);
      newState = [...selectedNames, newSelection];
    }

    let utter: any;

    if ('SpeechSynthesisUtterance' in window) {
      utter = new SpeechSynthesisUtterance(newSelection);
      utter.lang = 'tr-TR';
    }

    setTimeout(() => {
      if (newSelection !== '') {
        setSelectedNames(newState);
        setNextPlayer(USER);
        if (synthRef.current) {
          synthRef.current.speak(utter);
        }
        setTimeLeft(8);
      } else {
        setGameStatus(FINISHED);
        setWinner(USER);
      }
    }, timeToWait * 1000);

    clearTimeout();
  } else {
    console.log('Machine can not find a name.');
    console.log('Game is gonna finish, winner is the user.');

    setTimeout(() => {
      setGameStatus(FINISHED);
      setWinner(USER);
    }, 8000);
  }
};

export default listenMachine;
