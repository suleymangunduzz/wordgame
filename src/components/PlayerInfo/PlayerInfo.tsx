import React, { FC } from 'react';

import './PlayerInfo.scss';

import { FINISHED, MACHINE } from '../../utils/constants';

type Props = {
  gameStatus: string;
  nextPlayer: string;
};

const PlayerInfo: FC<Props> = ({ gameStatus, nextPlayer }) => {
  if (gameStatus !== FINISHED) {
    return <div className="PlayerInfo">{`Oynama sırası: ${nextPlayer || MACHINE}`}</div>;
  }

  return <></>;
};

export default PlayerInfo;
