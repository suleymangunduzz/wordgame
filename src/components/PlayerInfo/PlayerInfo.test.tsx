import React from 'react';
import { shallow } from 'enzyme';
import { NOT_STARTED, FINISHED, STARTED, MACHINE, USER } from '../../utils/constants';
import PlayerInfo from './PlayerInfo';

it('renders correctly Player Info when game is finished with machine.', () => {
  const wrapper = shallow(<PlayerInfo gameStatus={FINISHED} nextPlayer={MACHINE} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Player Info when game is finished with user.', () => {
  const wrapper = shallow(<PlayerInfo gameStatus={FINISHED} nextPlayer={USER} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Player Info when game is playing with machine.', () => {
  const wrapper = shallow(<PlayerInfo gameStatus={STARTED} nextPlayer={MACHINE} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Player Info when game is playing with user.', () => {
  const wrapper = shallow(<PlayerInfo gameStatus={STARTED} nextPlayer={USER} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Player Info when game is not started yet and player is machine.', () => {
  const wrapper = shallow(<PlayerInfo gameStatus={NOT_STARTED} nextPlayer={MACHINE} />);

  expect(wrapper).toMatchSnapshot();
});

it('renders correctly Player Info when game is not started yet and player is user.', () => {
  const wrapper = shallow(<PlayerInfo gameStatus={NOT_STARTED} nextPlayer={USER} />);

  expect(wrapper).toMatchSnapshot();
});
