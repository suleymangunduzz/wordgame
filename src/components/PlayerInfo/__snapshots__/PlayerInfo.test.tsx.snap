// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Player Info when game is finished with machine. 1`] = `<Fragment />`;

exports[`renders correctly Player Info when game is finished with user. 1`] = `<Fragment />`;

exports[`renders correctly Player Info when game is not started yet and player is machine. 1`] = `
<div
  className="PlayerInfo"
>
  Oynama sırası: MACHINE
</div>
`;

exports[`renders correctly Player Info when game is not started yet and player is user. 1`] = `
<div
  className="PlayerInfo"
>
  Oynama sırası: USER
</div>
`;

exports[`renders correctly Player Info when game is playing with machine. 1`] = `
<div
  className="PlayerInfo"
>
  Oynama sırası: MACHINE
</div>
`;

exports[`renders correctly Player Info when game is playing with user. 1`] = `
<div
  className="PlayerInfo"
>
  Oynama sırası: USER
</div>
`;
